### DISCUSSION TITLE
How to fly several CF using a Craziradio PA?

### DISCUSSION THREAD

**Discussion by Wallat1** [bitcraze/discussions#379] [Q&A]
Hi.

I would like to know if you have any method to fly several CF using a Craziradio PA?. I clarify that I have seen the examples of swarm, but I think that this does not work with the ramp.py program or if I am wrong, I would appreciate an explanation of how to use it.

Regards.

**Comment by krichardsson** [comment]
There are multiple ways of doing this, unfortunately there is not a lot of documentation so reading the examples and the source code is the way to find out how it works. I will try to add some pointers for you to get started.

The first thing to understand is that when talking to a Crazyflie all communication is asynchronous, that after sending a message to the Crazyflie your code will continue to execute, and at some later point in time you get a response back through a callback. 
This mode of communication makes sense in some cases, but in some other applications it is easier to use synchronous communication - you send a message and your function does not return until you get a response from the Crazyflie. There is a class called `SyncCrazyflie` that wraps the basic Crazyflie connection and provides a synchronous API. Both methods are used in the examples, the ramp.py is asynchronous for instance. 

You can connect to multiple Crazyflies asynchronously as well, but it is a bit messy and I don't think there are any examples for this. To make it easier the Swarm class wraps the complexity and creates one `SyncCrazyflie` instance, running in a separate thread per Crazyflie connection. The use of the Swarm class is still a bit complex but it hopefully makes it a bit easier.

Again checking out the examples is probably the best way to go.

To run the ramp.py examples using the Swarm class should not be too hard. Essentially call _ramp_motors() using swarm.parallel_safe().

See #326 for some more information on what channels/addresses to use on the different Crazyflies.

**Reply by Wallat1** [reply]
Hi 
Thank you so much for the help.

For simplicity I made the following program:

```
import logging
import time
from cflib.crazyflie.swarm import CachedCfFactory
from cflib.crazyflie.swarm import Swarm
from threading import Thread
import cflib
from cflib.crazyflie import Crazyflie

uris = {
    'radio://0/80/2M/E7E7E7E701',
    'radio://0/80/2M/E7E7E7E702',
}

def CFtest()
```

    cf1 = Crazyflie(rw_cache='./cache')
    cf2 = Crazyflie(rw_cache='./cache')
        
    cf1.commander.send_setpoint(0, 0, 0, 0)
    cf2.commander.send_setpoint(0, 0, 0, 0)

    cf1.commander.send_setpoint(0, 0, 0, 40000)
    cf2.commander.send_setpoint(0, 0, 0, 40000)
    time.sleep(0.5)
    cf1.commander.send_setpoint(0, 0, 0, 0)
    cf2.commander.send_setpoint(0, 0, 0, 0)
    time.sleep(0.1)
    cf1.close_link()
    cf2.close_link()

`if __name__ == '__main__':`
   
    cflib.crtp.init_drivers()
    factory = CachedCfFactory(rw_cache='./cache')
    with Swarm(uris, factory=factory) as swarm:
    swarm.parallel_safe(CFtest)

to try to use two CF with a crazyradio PA, however when I run it it tells me that there are problems with the threads, but I don't know exactly what it means, so could you please help me to solve my problem?.

Regards.

**Reply by krichardsson** [reply]
It looks like there are a few missundestandings here:
The Swarm class is [instantiating the Crazyflie objects](https://github.com/bitcraze/crazyflie-lib-python/blob/master/cflib/crazyflie/swarm.py#L82-L83) and this should not be done in your `CFtest()` function. Further more it is also connecting to the Crazyflies.

The connected SyncCrazyflie instance is [passed to the function, as can be seen in this example](https://github.com/bitcraze/crazyflie-lib-python/blob/master/examples/swarm/hl-commander-swarm.py#L40-L41). Your `CFtest()` function will be called for each Crazyflie (uri).

To replicate the behavior I think you are aiming for, it should look something like this (the setpoints look pretty weird though):
``` python
def CFtest(scf):
    # Get the cf object from the sync crazyflie
    cf = scf.cf
    # Call the set point functions, like in your code
    cf.commander.send_setpoint(0, 0, 0, 0)
    cf.commander.send_setpoint(0, 0, 0, 40000)
    time.sleep(0.5)
    cf.commander.send_setpoint(0, 0, 0, 0)
    time.sleep(0.1)
    # No closing of the link here, this is done by the swarm class when it goes out of scope
```

**Comment by knmcguire** [comment]
We have an example for controlling a swarm with the crazyflie python library here: https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/user-guides/sbs_swarm_interface/

Also we advise you to look at the Crazyswarm project:
ROS1: https://crazyswarm.readthedocs.io/en/latest/
ROS2: https://imrclab.github.io/crazyswarm2/

**Reply by Wallat1** [reply]
Hi.

Thank you so much for the help, it was very useful. 

However I hope and you can help me solve my problem, since currently I am using 4 crazyflies and a crazyradio for each one, which is not very efficient if I need to use a greater number of CF and also I have the problem that they do not take off at the same time, which I attribute to the use of the 4 crazyradio PA.

Regards.

**Reply by knmcguire** [reply]
The first tutorial shows how to connect one crazyradio for all 4, so that should work already. Also the crazyswarm projects all work with one crazyradio for multiple crazyflies.

The URI's first number indicates which Crazyradio to use for the crazyflie: 'radio://0/20/2M/E7E7E7E701'. Either you use 'radio://1/20/2M/E7E7E7E702' for the second drone (uses a second crazyradio), or 'radio://0/20/2M/E7E7E7E702' which uses the same radio as before. For the latter to work, it is important to use the swarm class as indicated in the first tutorial.

**Reply by Wallat1** [reply]
Hi.

I was previously working with crazyswarm but cmdVel() which is equivalent to send_setpoint() does not work and the forum managers have not answered for a long time, for this reason I am using crazyflie lib python.

Regards.

**Reply by Wallat1** [reply]
Hi.

I have followed the instructions of the tutorial, but I heve yet doubt about how to use send_setpoint() with the swarm, for such motive I would like to know, if do you have some tutorial where  send_setpoint() is used with swarm?.

Regards.
